panic during shutdown flow

good instance:
```
[Sarama] 2018/03/15 17:16:48 kafka: error while consuming m-12574-M-ops/27: kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
[Sarama] 2018/03/15 17:16:48 consumer/m-12574-M-ops/27 shutting down because kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
2018/03/15 17:16:48 [kafkamdm.go:310 consumePartition()] [E] kafka-mdm: kafka consumer for m-12574-M-ops:27 has shutdown. stop consuming
2018/03/15 17:16:48 [I] An input plugin signalled a fatal error. Shutting down
2018/03/15 17:16:48 [I] CLU manager: HTTPNode mt-read03-12574-medium-ops-b-1218431600-1d3fz has left the cluster
2018/03/15 17:16:48 [I] CLU manager: HTTPNode mt-read01-12574-medium-ops-b-3464612116-st4js has left the cluster
2018/03/15 17:16:48 [I] API shutdown started.
2018/03/15 17:16:49 [I] API accept tcp [::]:6060: use of closed network connection
2018/03/15 17:16:49 [I] Shutting down kafka-mdm consumer
[Sarama] 2018/03/15 17:16:49 consumer/broker/786437 closed dead subscription to m-12574-M-ops/31
[Sarama] 2018/03/15 17:16:49 consumer/broker/786439 closed dead subscription to m-12574-M-ops/25
[Sarama] 2018/03/15 17:16:49 consumer/broker/786439 closed dead subscription to m-12574-M-ops/29
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:31 ended.
[Sarama] 2018/03/15 17:16:49 consumer/broker/786438 closed dead subscription to m-12574-M-ops/24
[Sarama] 2018/03/15 17:16:49 consumer/broker/786438 closed dead subscription to m-12574-M-ops/28
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:29 ended.
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:25 ended.
[Sarama] 2018/03/15 17:16:49 consumer/broker/786440 closed dead subscription to m-12574-M-ops/26
[Sarama] 2018/03/15 17:16:49 consumer/broker/786440 closed dead subscription to m-12574-M-ops/30
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:28 ended.
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:24 ended.
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:30 ended.
2018/03/15 17:16:49 [I] kafka-mdm consumer for m-12574-M-ops:26 ended.
2018/03/15 17:16:49 [I] Closing partitionsOffset DB.
[Sarama] 2018/03/15 17:16:49 Closing Client
[Sarama] 2018/03/15 17:16:49 Closed connection to broker kafka:9092
[Sarama] 2018/03/15 17:16:49 Closed connection to broker 10.12.0.7:9092
[Sarama] 2018/03/15 17:16:49 Closed connection to broker 10.12.0.6:9092
[Sarama] 2018/03/15 17:16:49 Closed connection to broker 10.12.0.5:9092
2018/03/15 17:16:49 [I] kafka-mdm consumer finished shutdown
2018/03/15 17:16:49 [I] closing store
[Sarama] 2018/03/15 17:16:49 Closed connection to broker 10.12.0.8:9092
2018/03/15 17:16:49 [I] cassandra-idx stopping
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] cassandra-idx writeQueue handler ended.
2018/03/15 17:16:49 [I] terminating.
```

bad instance:
```
[Sarama] 2018/03/15 17:16:48 kafka: error while consuming m-12574-M-ops/23: kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
[Sarama] 2018/03/15 17:16:48 consumer/m-12574-M-ops/23 shutting down because kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
[Sarama] 2018/03/15 17:16:48 kafka: error while consuming m-12574-M-ops/19: kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
[Sarama] 2018/03/15 17:16:48 consumer/m-12574-M-ops/19 shutting down because kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition.
2018/03/15 17:16:48 [kafkamdm.go:310 consumePartition()] [E] kafka-mdm: kafka consumer for m-12574-M-ops:19 has shutdown. stop consuming
2018/03/15 17:16:48 [I] An input plugin signalled a fatal error. Shutting down
2018/03/15 17:16:48 [I] CLU manager: HTTPNode mt-read02-12574-medium-ops-b-1100597842-p69pz has left the cluster
2018/03/15 17:16:48 [kafkamdm.go:310 consumePartition()] [E] kafka-mdm: kafka consumer for m-12574-M-ops:23 has shutdown. stop consuming
panic: close of closed channel

goroutine 1264 [running]:
github.com/grafana/metrictank/input/kafkamdm.(*KafkaMdm).consumePartition(0xc4208d65a0, 0xc420034f47, 0xd, 0x300000017, 0x3baa4f7e7)
	/go/src/github.com/grafana/metrictank/input/kafkamdm/kafkamdm.go:314 +0x9bc
created by github.com/grafana/metrictank/input/kafkamdm.(*KafkaMdm).Start
	/go/src/github.com/grafana/metrictank/input/kafkamdm/kafkamdm.go:225 +0x106

```

--- a/cmd/metrictank/metrictank.go
+++ b/cmd/metrictank/metrictank.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"context"
 	"flag"
 	"fmt"
 	l "log"
@@ -402,12 +403,12 @@ func main() {
 	/***********************************
 		Start our inputs
 	***********************************/
-	pluginFatal := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
 	for _, plugin := range inputs {
 		if carbonPlugin, ok := plugin.(*inCarbon.Carbon); ok {
 			carbonPlugin.IntervalGetter(inCarbon.NewIndexIntervalGetter(metricIndex))
 		}
-		err = plugin.Start(input.NewDefaultHandler(metrics, metricIndex, plugin.Name()), pluginFatal)
+		err = plugin.Start(input.NewDefaultHandler(metrics, metricIndex, plugin.Name()), cancel)
 		if err != nil {
 			shutdown()
 			return
@@ -439,7 +440,7 @@ func main() {
 	select {
 	case sig := <-sigChan:
 		log.Infof("Received signal %q. Shutting down", sig)
-	case <-pluginFatal:
+	case <-ctx.Done():
 		log.Info("An input plugin signalled a fatal error. Shutting down")
 	}
 	shutdown()

--- a/cmd/mt-kafka-mdm-sniff-out-of-order/main.go
+++ b/cmd/mt-kafka-mdm-sniff-out-of-order/main.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"context"
 	"flag"
 	"fmt"
 	"math/rand"
@@ -211,14 +212,14 @@ func main() {
 	stats.NewDevnull() // make sure metrics don't pile up without getting discarded
 
 	mdm := inKafkaMdm.New()
-	pluginFatal := make(chan struct{})
-	mdm.Start(newInputOOOFinder(*format), pluginFatal)
+	ctx, cancel := context.WithCancel(context.Background())
+	mdm.Start(newInputOOOFinder(*format), cancel)
 	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
 	select {
 	case sig := <-sigChan:
 		log.Infof("Received signal %q. Shutting down", sig)
-	case <-pluginFatal:
+	case <-ctx.Done():
 		log.Info("Mdm input plugin signalled a fatal error. Shutting down")
 	}
 	mdm.Stop()

--- a/cmd/mt-kafka-mdm-sniff/main.go
+++ b/cmd/mt-kafka-mdm-sniff/main.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"context"
 	"flag"
 	"fmt"
 	"math/rand"
@@ -158,14 +159,14 @@ func main() {
 	stats.NewDevnull() // make sure metrics don't pile up without getting discarded
 
 	mdm := inKafkaMdm.New()
-	pluginFatal := make(chan struct{})
-	mdm.Start(newInputPrinter(*formatMd, *formatP), pluginFatal)
+	ctx, cancel := context.WithCancel(context.Background())
+	mdm.Start(newInputPrinter(*formatMd, *formatP), cancel)
 	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
 	select {
 	case sig := <-sigChan:
 		log.Infof("Received signal %q. Shutting down", sig)
-	case <-pluginFatal:
+	case <-ctx.Done():
 		log.Info("Mdm input plugin signalled a fatal error. Shutting down")
 	}
 	mdm.Stop()

--- a/input/carbon/carbon.go
+++ b/input/carbon/carbon.go
@@ -4,6 +4,7 @@ package carbon
 
 import (
 	"bufio"
+	"context"
 	"flag"
 	"io"
 	"net"
@@ -106,7 +107,7 @@ func (c *Carbon) IntervalGetter(i IntervalGetter) {
 	c.intervalGetter = i
 }
 
-func (c *Carbon) Start(handler input.Handler, fatal chan struct{}) error {
+func (c *Carbon) Start(handler input.Handler, cancel context.CancelFunc) error {
 	c.Handler = handler
 	l, err := net.ListenTCP("tcp", c.addr)
 	if nil != err {

--- a/input/kafkamdm/kafkamdm.go
+++ b/input/kafkamdm/kafkamdm.go
@@ -1,6 +1,7 @@
 package kafkamdm
 
 import (
+	"context"
 	"flag"
 	"fmt"
 	"strconv"
@@ -37,7 +38,7 @@ type KafkaMdm struct {
 	// signal to PartitionConsumers to shutdown
 	stopConsuming chan struct{}
 	// signal to caller that it should shutdown
-	fatal chan struct{}
+	cancel context.CancelFunc
 }
 
 func (k *KafkaMdm) Name() string {
@@ -209,9 +210,9 @@ func New() *KafkaMdm {
 	return &k
 }
 
-func (k *KafkaMdm) Start(handler input.Handler, fatal chan struct{}) error {
+func (k *KafkaMdm) Start(handler input.Handler, cancel context.CancelFunc) error {
 	k.Handler = handler
-	k.fatal = fatal
+	k.cancel = cancel
 	var err error
 	for _, topic := range topics {
 		for _, partition := range partitions {
@@ -288,7 +289,7 @@ func (k *KafkaMdm) consumePartition(topic string, partition int32, currentOffset
 	newest, err := k.tryGetOffset(topic, partition, sarama.OffsetNewest, 7, time.Second*10)
 	if err != nil {
 		log.Errorf("kafkamdm: %s", err.Error())
-		close(k.fatal)
+		k.cancel()
 		return
 	}
 	if currentOffset == sarama.OffsetNewest {
@@ -297,7 +298,7 @@ func (k *KafkaMdm) consumePartition(topic string, partition int32, currentOffset
 		currentOffset, err = k.tryGetOffset(topic, partition, sarama.OffsetOldest, 7, time.Second*10)
 		if err != nil {
 			log.Errorf("kafkamdm: %s", err.Error())
-			close(k.fatal)
+			k.cancel()
 			return
 		}
 	}
@@ -310,7 +311,7 @@ func (k *KafkaMdm) consumePartition(topic string, partition int32, currentOffset
 	pc, err := k.consumer.ConsumePartition(topic, partition, currentOffset)
 	if err != nil {
 		log.Errorf("kafkamdm: failed to start partitionConsumer for %s:%d. %s", topic, partition, err)
-		close(k.fatal)
+		k.cancel()
 		return
 	}
 	messages := pc.Messages()
@@ -324,7 +325,7 @@ func (k *KafkaMdm) consumePartition(topic string, partition int32, currentOffset
 				if err := offsetMgr.Commit(topic, partition, currentOffset); err != nil {
 					log.Errorf("kafkamdm: failed to commit offset for %s:%d, %s", topic, partition, err)
 				}
-				close(k.fatal)
+				k.cancel()
 				return
 			}
 			log.Debugf("kafkamdm: received message: Topic %s, Partition: %d, Offset: %d, Key: %x", msg.Topic, msg.Partition, msg.Offset, msg.Key)

--- a/input/plugin.go
+++ b/input/plugin.go
@@ -1,13 +1,15 @@
 package input
 
+import "context"
+
 type Plugin interface {
 	Name() string
 	// Start starts the plugin.
-	// The plugin closes the fatal chan should any non-recoverable error occur after Start has returned.
-	// if Start returns an error, or the fatal chan is closed by the plugin,
+	// The plugin calls cancel should any non-recoverable error occur after Start has returned.
+	// if Start returns an error, or cancel is called by the plugin,
 	// the caller (e.g. main process) should shut down all its resources and exit.
 	// Note that upon fatal close, metrictank will call Stop() on all plugins, also the one that triggered it.
-	Start(handler Handler, fatal chan struct{}) error
+	Start(handler Handler, cancel context.CancelFunc) error
 	MaintainPriority()
 	ExplainPriority() interface{}
 	Stop() // Should block until shutdown is complete.

--- a/input/prometheus/prometheus.go
+++ b/input/prometheus/prometheus.go
@@ -1,6 +1,7 @@
 package prometheus
 
 import (
+	"context"
 	"flag"
 	"fmt"
 	"io/ioutil"
@@ -25,7 +26,6 @@ var (
 
 type prometheusWriteHandler struct {
 	input.Handler
-	quit chan struct{}
 }
 
 func New() *prometheusWriteHandler {
@@ -36,9 +36,8 @@ func (p *prometheusWriteHandler) Name() string {
 	return "prometheus"
 }
 
-func (p *prometheusWriteHandler) Start(handler input.Handler, fatal chan struct{}) error {
+func (p *prometheusWriteHandler) Start(handler input.Handler, cancel context.CancelFunc) error {
 	p.Handler = handler
-	p.quit = fatal
 	ConfigSetup()
 
 	mux := http.NewServeMux()

